File: scrudbeans-api/src/main/java/com/github/manosbatsis/scrudbeans/api/mdd/registry/IdentifierAdaptersRegistry.java
package com.github.manosbatsis.scrudbeans.api.mdd.registry;

import com.github.manosbatsis.scrudbeans.api.mdd.model.IdentifierAdapter;
import org.springframework.util.Assert;

import java.util.HashMap;

public class IdentifierAdaptersRegistry {

    protected static final HashMap<String, IdentifierAdapter> adaptersMap = new HashMap<>();

    /**
     * Register an appropriate predicate factory for the given class
     */
    public static void addAdapterForClass(Class clazz, IdentifierAdapter adapter) {
        Assert.notNull(clazz, "clazz cannot be null");
        Assert.notNull(adapter, "adapter cannot be null");
        adaptersMap.put(clazz.getCanonicalName(), adapter);
    }

    /**
     * Register an appropriate predicate factory for the given class
     */
    public static void addAdapterForClass(Class clazz, Class<IdentifierAdapter> adapterClass) {
        Assert.notNull(clazz, "clazz cannot be null");
        Assert.notNull(adapterClass, "adapterClass cannot be null");
        try {
            adaptersMap.put(clazz.getCanonicalName(), adapterClass.newInstance());
        } catch (Exception e) {
            throw new RuntimeException("Failed creating identifier adapter instance", e);
        }
    }

    /**
     * Get an appropriate {@link IdentifierAdapter} factory for the given class
     */
    public static <T> IdentifierAdapter<T, ?> getAdapterForClass(Class<T> clazz) {
        return adaptersMap.get(clazz.getCanonicalName());
    }

    /**
     * Get an appropriate {@link IdentifierAdapter} factory for the given class
     */
    public static boolean containsAdapterForClass(Class<?> clazz) {
        return getAdapterForClass(clazz) != null;
    }
}

File: scrudbeans-api/src/main/kotlin/com/github/manosbatsis/scrudbeans/api/mdd/model/IdentifierAdaptersRegistry.kt
package com.github.manosbatsis.scrudbeans.api.mdd.model

import org.springframework.util.Assert

object IdentifierAdaptersRegistry {
    internal val adaptersMap = HashMap<String, IdentifierAdapter<*, *>>()

    /**
     * Register an appropriate predicate factory for the given class
     */
    fun addAdapterForClass(clazz: Class<*>, adapter: IdentifierAdapter<*, *>) {
        Assert.notNull(clazz, "clazz cannot be null")
        Assert.notNull(adapter, "adapter cannot be null")
        adaptersMap[clazz.canonicalName] = adapter
    }

    /**
     * Register an appropriate predicate factory for the given class
     */
    fun addAdapterForClass(clazz: Class<*>, adapterClass: Class<IdentifierAdapter<*, *>?>) {
        Assert.notNull(clazz, "clazz cannot be null")
        Assert.notNull(adapterClass, "adapterClass cannot be null")
        try {
            adaptersMap[clazz.canonicalName] = adapterClass.newInstance() as IdentifierAdapter<*, *>
        } catch (e: Exception) {
            throw RuntimeException("Failed creating identifier adapter instance", e)
        }
    }

    /**
     * Get an appropriate [IdentifierAdapter] factory for the given class
     */
    @JvmStatic
    fun <T> getAdapterForClass(clazz: Class<T>): IdentifierAdapter<T, *>? {
        return adaptersMap[clazz.canonicalName] as IdentifierAdapter<T, *>
    }

    /**
     * Get an appropriate [IdentifierAdapter] factory for the given class
     */
    fun containsAdapterForClass(clazz: Class<*>): Boolean {
        return getAdapterForClass(clazz) != null
    }
}